### PR TITLE
Add support for D interface files

### PR DESCRIPTION
**Pull request by Alexibu**
D compilers can all output `.di` files which are source code files with all the implementation of non template code removed. This work adds a cross compiler method of producing these files as secondary outputs when compiling an object file from a source file.

https://dlang.org/dmd-linux.html#interface-files

* [x] I have created a new test or updated the unit tests to cover the new/changed functionality.
* [x] I have updated `CHANGES.txt` (and read the `README.rst`)
* [x] I have updated the appropriate documentation

### CODE REVIEW

File: test/D/di/Common/__init__.py
Comment (bdbaddog) [inline test/D/di/Common/__init__.py:1]: No __init__ is needed in the test tree..

**Comment by bdbaddog** [comment]
You'll need to add to the doc files for the new variables.
ldc.xml, gdc.xml, dmd.xml, DCommon.xml (start here).
You should be able to copy/paste from other variables in each.

Also can you update your copyright blocks to match the new templates we have for such?
See: https://github.com/SCons/scons/blob/master/template/Tests.py

**Comment by bdbaddog** [comment]
Please add a blurb to RELEASE.txt as well. (You can likely just copy what you have in CHANGES.txt into the appropriate section in RELEASE.txt)

File: SCons/Tool/DCommon.xml [hunk 1]
```diff
@@ -350,6 +350,45 @@ See also &cv-link-DLINKFLAGS; for linking static objects.
...
 
+<cvar name="DIFILESUFFIX">
+<summary>
+<para>
+	Suffix of d include files default is .di
+</para>
+</summary>
+</cvar>
+
+<cvar name="DIFILESUFFIX">
```
Comment (bdbaddog) [inline SCons/Tool/DCommon.xml]: duplicate

File: SCons/Tool/DCommon.xml [hunk 2]
```diff
@@ -350,6 +350,37 @@ See also &cv-link-DLINKFLAGS; for linking static objects.
...
 
+<cvar name="DIFILESUFFIX">
+<summary>
+<para>
+Suffix of d include files default is .di
+</para>
+</summary>
+</cvar>
+
+<cvar name="DINTFDIRKEY">
```
Comment (bdbaddog) [inline SCons/Tool/DCommon.xml]: DINTFDIRKEY ? seems an odd variable name.
Can you explain?

File: SCons/Tool/dmd.py [hunk 3]
```diff
@@ -87,18 +87,20 @@ def generate(env) -> None:
...
+    static_obj.add_emitter('.d', DCommon.DStaticObjectEmitter)
+    shared_obj.add_emitter('.d', DCommon.DSharedObjectEmitter)
 
     env['DC'] = env.Detect(['dmd', 'ldmd2', 'gdmd']) or 'dmd'
-    env['DCOM'] = '$DC $_DINCFLAGS $_DVERFLAGS $_DDEBUGFLAGS $_DFLAGS -c -of$TARGET $SOURCES'
+    env['DCOM'] = '$DC $_DINCFLAGS $_DVERFLAGS $_DDEBUGFLAGS $_DFLAGS $_DINTFDIR -c -of$TARGET $SOURCES'
     env['_DINCFLAGS'] = '${_concat(DINCPREFIX, DPATH, DINCSUFFIX, __env__, RDirs, TARGET, SOURCE)}'
     env['_DVERFLAGS'] = '${_concat(DVERPREFIX, DVERSIONS, DVERSUFFIX, __env__)}'
     env['_DDEBUGFLAGS'] = '${_concat(DDEBUGPREFIX, DDEBUG, DDEBUGSUFFIX, __env__)}'
+    env['_DINTFDIR'] = '${_optWithIxes(DINTFDIRPREFIX, DINTFDIRKEY, DINTFDIRSUFFIX, __env__, Dirs, TARGET, SOURCE)}'
```
Comment (bdbaddog) [inline SCons/Tool/dmd.py]: Can you explain why the normal _concat wouldn't' do here?

File: SCons/Tool/gdc.py [hunk 4]
```diff
@@ -82,13 +84,17 @@ def generate(env) -> None:
 
     env['DINCPREFIX'] = '-I'
     env['DINCSUFFIX'] = ''
-    env['DVERPREFIX'] = '-version='
+    env['DVERPREFIX'] = '-fversion='
     env['DVERSUFFIX'] = ''
-    env['DDEBUGPREFIX'] = '-debug='
+    env['DDEBUGPREFIX'] = '-fdebug='
```
Comment (bdbaddog) [inline SCons/Tool/gdc.py:88]: was the previous value always wrong? Or did this change at some version of gdc?

File: SCons/Tool/DCommon.xml (hunk 2 above)
Comment (Alexibu) [reply SCons/Tool/DCommon.xml]: Can't explain, wrote this a long time ago, merging now because I found that my code had been broken.

Comment (Alexibu) [reply SCons/Tool/DCommon.xml]: DINTFDIR ? Don't need the key

File: SCons/Tool/gdc.py (hunk 4 above)
Comment (Alexibu) [reply SCons/Tool/gdc.py:88]: Always wrong.

Comment (Alexibu) [reply SCons/Tool/gdc.py:88]: There is a gdmd wrapper for gdc that has the same arguments as dmd, but the arguments are different here.

File: SCons/Tool/dmd.py (hunk 3 above)
Comment (Alexibu) [reply SCons/Tool/dmd.py]: No not at the moment, I'll test it with concat, thanks.

Comment (Alexibu) [reply SCons/Tool/dmd.py]: Made it work with concat, although there should only be one DINTFDIR unlike DFLAGS etc.

File: test/D/di/Image/SConstruct_template [hunk 5]
```diff
@@ -0,0 +1,15 @@
+# -*- mode:python; coding:utf-8; -*-
+
+import os
+import SCons.Node
+
+environment = Environment(
+    tools=['link','ar' ,'{}']
+)
+
+environment.Decider("content")
```
Comment (bdbaddog) [inline test/D/di/Image/SConstruct_template]: Not necessary, that's the default.

File: test/D/di/Image/SConstruct_template [hunk 6]
```diff
@@ -0,0 +1,15 @@
+# -*- mode:python; coding:utf-8; -*-
+
+import os
+import SCons.Node
```
Comment (bdbaddog) [inline test/D/di/Image/SConstruct_template]: This import's also not needed.

File: test/D/di/Image/SConstruct_template [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+
```
Comment (bdbaddog) [inline test/D/di/Image/SConstruct_template:4]: Please add the following. It instructs SCons to not initialize any tools in the DefaultEnvironment(), which will speed up the test (especially on windows)

`DefaultEnvironment(tools=[])`

**Comment by bdbaddog** [comment]
Please remove all of these, we no longer use them.

`__revision__ = "__FILE__ __REVISION__ __DATE__ __DEVELOPER__"`

**Comment by bdbaddog** [comment]
So is `DINTFDIR` meant to only be a single directory and not a path?
So you really don't need `_concat()` then.

You should be able to use something like this. I think it's a bit simpler?

```
# Here's what it should expand to
env['_DINTFDIR_EXPANDED'] = '${DINTFDIRPREFIX}${DINTFDIR}${DINTFDIRSUFFIX}'
# Only expand it if DINTFDIR is specified
env['_DINTFDIR']='${DINTFDIR and _DINTFDIR_EXPANDED}'
```

File: SCons/Tool/DCommon.xml (hunk 2 above)
Comment (mwichmann) [reply SCons/Tool/DCommon.xml]: I'm presuming it's for this:
```
       -Hd dir
           Same as -H, but writes interface files to directory dir.  This
           option can be used with -Hf file to independently set the output
           file and directory path.
```
Is there a better name we could pick, then?

**Comment by bdbaddog** [comment]
If you wouldn't mind. (And you've set up the PR this way).
The maintainer's can help bring this PR to completion.

Or we can PR your branch with changes.
Whichever is better for you @Alexibu

File: SCons/Tool/DCommon.xml (hunk 2 above)
Comment (mwichmann) [reply SCons/Tool/DCommon.xml]: `DIFILEDIR`?

**Comment by Alexibu** [comment]
I am sure the maintainers could do a better job than me, and would be happy for them to help.

File: SCons/Tool/DCommon.xml (hunk 2 above)
Comment (Alexibu) [reply SCons/Tool/DCommon.xml]: Yes it's for -Hd dir. The file specific -Hf version should probably have a SCons implementation, but I haven't used it.
DIFILEDIR also good.

**Comment by Alexibu** [comment]
```
# Here's what it should expand to
env['_DINTFDIR_EXPANDED'] = '${DINTFDIRPREFIX}${DINTFDIR}${DINTFDIRSUFFIX}'
# Only expand it if DINTFDIR is specified
env['_DINTFDIR']='${DINTFDIR and _DINTFDIR_EXPANDED}'
```

I couldn't get this to work.
Maintainers should also check that the custom two target emitter in DCommon.py that I wrote is the right thing to do.

File: SCons/Tool/DCommon.xml (hunk 2 above)
Comment (bdbaddog) [reply SCons/Tool/DCommon.xml]: How about `DI_FILE_DIR` ?  @mwichmann 
I know we have a long history of VERYLONGVARIABLESWHICHAREHARDTOREAD... but I think there's no benefit in continuing that?

**Comment by bdbaddog** [comment]
@Alexibu a couple questions as I look at making some changes to your PR

It looks like a D compile will only find the `.di` files if they are in a path which is included in `DPATH`
So this works:
```
$ /home/bdeegan/.virtualenvs/scons_311/bin/python /home/bdeegan/devel/scons/git/users/alexibu/scripts/scons.py
scons: Reading SConscript files ...
scons: done reading SConscript files.
scons: Building targets ...
dmd -Iinclude -Hd=include -c -ofhelloWorldMain.o helloWorldMain.d
dmd -Iinclude -Hd=include -c -ofsource/helloWorld.o source/helloWorld.d
dmd -ofhw source/helloWorld.o helloWorldMain.o
scons: done building targets.
```

But this fails
```
$ /home/bdeegan/.virtualenvs/scons_311/bin/python /home/bdeegan/devel/scons/git/users/alexibu/scripts/scons.py
scons: Reading SConscript files ...
scons: done reading SConscript files.
scons: Building targets ...
dmd -I. -Hd=include -c -ofhelloWorldMain.o helloWorldMain.d
helloWorldMain.d(1): Error: module `helloWorld` is in file 'helloWorld.d' which cannot be read
import path[0] = .
import path[1] = /usr/include/dmd/phobos
import path[2] = /usr/include/dmd/druntime/import
scons: *** [helloWorldMain.o] Error 1
scons: building terminated because of errors.
```

Since the .di file is the include dir and that's not specified with `-Iinclude` 

Does that seem correct?

**Comment by bdbaddog** [comment]
Additionally it looks like (at least for dmd), you can specify either a directory for the .di file to be written to, or the specific file you want written?

`-Hd include`
or
`-Hf include/helloWorld.di`

For example.
Is that correct and supported by dmd, ldc, gdc ?

**Comment by Alexibu** [comment]
> It looks like a D compile will only find the `.di` files if they are in a path which is included in `DPATH` So this works:

Yes thats right. The compiler searches for the di files in the same way as .d files. The compiler prefers .d files over .di files if both are available.
The -Hd arguments generate the .di files from the .d file being compiled, the -I argument is necessary to allow it to find the di files from other compilation steps.

**Comment by Alexibu** [comment]
> Additionally it looks like (at least for dmd), you can specify either a directory for the .di file to be written to, or the specific file you want written?

I believe they all have equivalent arguments for per file naming of generated di files, but I've never used it.

**Comment by bdbaddog** [comment]
@Alexibu - I filed a PR against your branch with my updates. Still a work in progress.

**Comment by bdbaddog** [comment]
> > It looks like a D compile will only find the `.di` files if they are in a path which is included in `DPATH` So this works:
> 
> Yes thats right. The compiler searches for the di files in the same way as .d files. The compiler prefers .d files over .di files if both are available. The -Hd arguments generate the .di files from the .d file being compiled, the -I argument is necessary to allow it to find the di files from other compilation steps.

Did you mean the inverse?
From dmd docs

> When the compiler resolves an import declaration, it first looks for a .di D interface file, then it looks for a D source file.

**Comment by Alexibu** [comment]
> Did you mean the inverse? From dmd docs
> 
> > When the compiler resolves an import declaration, it first looks for a .di D interface file, then it looks for a D source file.

No I was not aware of the docs, I remember having to take special measures to prevent the compiler seeing the .d files to force it to use di files. I could have been mistaken. In my projects that use di files, there is also the use of scons build dirs, which adds additional complexity.

**Comment by bdbaddog** [comment]
@Alexibu - please merge https://github.com/Alexibu/scons/pull/1 unless you see some issues with it.
As far as I can tell .di files should no work properly. There should be no need to use Depends() as well.
Take a look at the SConstruct_template from the tests. I've modified it to be as simple as possible and still work.. ;)

**Pull request merged**